harness/determined/pytorch/deepspeed/_deepspeed_context.py

```diff
@@ -0,0 +1,347 @@
+                "Please make sure wrap_mpu is only called once in the trial definition."
+            )
+        self._called_wrap_mpu = True
+        old_mpu = self._mpu
+        self._mpu = mpu
+        if old_mpu.get_data_parallel_world_size() != mpu.get_data_parallel_world_size():
+            (
+                self.env._per_slot_batch_size,
+                self.env._global_batch_size,
+            ) = self._calculate_batch_sizes()
+            logging.warning(
+                "Wrapped MPU uses model parallelism.  Changing per slot batch size "
+                f"to {self.get_per_slot_batch_size()} and "
+                f"global batch size to {self.get_global_batch_size()}"
+            )
+        return mpu
+
+    @property
+    def mpu(self) -> det_ds.ModelParallelUnit:
+        return self._mpu
+
+    def wrap_model_engine(self, model: torch.nn.Module) -> torch.nn.Module:
+        """Returns a wrapped model engine.
+
+        In the background, we perform checks to properly handle pipeline parallelism if
+        the model engine is a PipelineEngine.  We also recompute batch sizes to match
+        the deepspeed config.
+        """
+
+        if self.env.managed_training:
+            model = model.to(self.device)
+
+        # Pipeline parallel model engine has its own MPU that we will use here.
+        if isinstance(model, deepspeed.PipelineEngine):
+            self._use_pipeline_parallel = True
+            self._mpu = det_ds.DeepSpeedMPU(model.mpu)
+
+        recompute_batch_size = False
+
+        # Check to make sure that Determined's global batch size matches the model engine's.
+        # If not, overwrite Determined's batch size.
+        if self.get_global_batch_size() != model.train_batch_size():
+            logging.warning(
+                f"Setting global batch size to {model.train_batch_size()} to match the "
+                "deepspeed config.  To prevent this from happening, you can call "
+                "self.context.overwrite_deepspeed_config(base_ds_config) to get a consistent"
+                "deepspeed config dict which you can pass to the config field when calling"
+                "deepspeed.initialize to build the model engine."
+            )
+            self.env._global_batch_size = model.train_batch_size()
+            recompute_batch_size = True
+
+        if self._train_micro_batch_size_per_gpu is None:
+            self._train_micro_batch_size_per_gpu = int(model.train_micro_batch_size_per_gpu())
+            recompute_batch_size = True
+        # If multiple model engines are wrapped, we will make sure that they have the same
+        # micro batch size.
+        assert (
+            self._train_micro_batch_size_per_gpu == model.train_micro_batch_size_per_gpu()
+        ), "micro batch size do not match across DeepSpeed model engines."
+
+        if recompute_batch_size:
+            (
+                self.env._per_slot_batch_size,
+                self.env._global_batch_size,
```

Comment:
This `EnvContext` object is terrible.  It's similar to the ClusterInfo in its purpose, but was used in a very bad way; rather than break up the One Big Config object at the top layers of code and pass specific necessary values into lower layers, we just passed the One Big Config down to the lowest layers of code and called it a day.  Then writing unit tests was hard because we had to construct a Mock One Big Config object to test almost anything.

Anyway, one thing we never did was modify the `env` at runtime.  Let's not start now.  Keep a source of truth somewhere other than the legacy EnvContext object and modify that other source of truth instead.

Reply:
I'm going to use the DeepSpeedContext as the source of truth.  If we do not modify `env` then there will be discrepancy in the global batch size and per slot batch size in the DeepSpeedContext and env.  Is that okay?

Reply:
Hm... I guess I could go either way on this one.  You can decide what you like best.

Users should not be using `context.env` no matter what, IMO.